### PR TITLE
Add OAuth based authentication and JWT access tokens for the API (sto…

### PR DESCRIPTION
**Pull request by matyix**
…red in Vault)

### CODE REVIEW

**Comment by matyix** [comment]
@stoader @baluchicken @colin014 please review - auth now is switched off so does not affect the 0.2.0 release. However, due to the coming GKE PR and the `post 0.2.0` refactor would be good to have it on master asap. Rebase and merge if it's OK before we tag.

File: auth/authn.go [hunk 1]
```diff
@@ -67,50 +137,100 @@ func loadPublicKey(data []byte) (interface{}, error) {
...
-		c.Next()
-	})
+//Auth0Handler handler for Gin
+func Auth0Handler(c *gin.Context) {
+	currentUser := Auth.GetCurrentUser(c.Request)
+	if currentUser != nil {
+		return
+	}
+
+	accessToken, err := validator.ValidateRequest(c.Request)
```
Comment (baluchicken) [inline auth/authn.go]: I think we should factor this logic to a function.

File: auth/authn.go [hunk 2]
```diff
@@ -67,50 +137,100 @@ func loadPublicKey(data []byte) (interface{}, error) {
...
+	if err != nil {
+		cloud.SetResponseBodyJson(c, http.StatusUnauthorized, gin.H{
+			cloud.JsonKeyError: "invalid token",
+		})
+		c.Abort()
+		banzaiUtils.LogInfo(banzaiConstants.TagAuth, "Invalid token:", err)
+		return
+	}
+
+	claims := map[string]interface{}{}
```
Comment (baluchicken) [inline auth/authn.go]: I think we should factor this logic to a function.

File: auth/authn.go [hunk 3]
```diff
@@ -67,50 +137,100 @@ func loadPublicKey(data []byte) (interface{}, error) {
...
+			cloud.JsonKeyError: "invalid claims",
+		})
+		c.Abort()
+		banzaiUtils.LogInfo(banzaiConstants.TagAuth, "Invalid claims:", err)
+		return
+	}
+
+	userID := claims["sub"].(string)
+	tokenID := claims["jti"].(string)
+	isTokenValid, err := lookupAccessToken(userID, tokenID)
```
Comment (baluchicken) [inline auth/authn.go]: I think we should factor this logic to a function.

File: conf/config.toml.example [hunk 4]
```diff
@@ -29,3 +29,10 @@ retrySleepSeconds = 15
 
 # SSH key path
 #dev.keypath = ""
+
+# GitHub settings
+clientid = ""
```
Comment (baluchicken) [inline conf/config.toml.example]: Are those really a dev settings?

**Comment by matyix** [comment]
@stoader pls review requested changes

**Comment by matyix** [comment]
thanks @stoader @baluchicken

**Pull request merged**